File: 213_robber2.java
class Solution {
    private int r(int[] nums) {
        int a = 0, b = 0, ans = 0;
        for (int i = 0; i < nums.length; ++i) {
            ans = Math.max(b + nums[i], a);
            b = a;
            a = ans;
        }
        return ans;
    }
    public int rob(int[] nums) {
        if (nums.length == 1) { return nums[0]; }
        int[] t1 = Arrays.copyOfRange(nums, 0, nums.length - 1);
        int[] t2 = Arrays.copyOfRange(nums, 1, nums.length);
        return Math.max(r(t1), r(t2));
    }
}

File: 213_robber2.cpp
class Solution {
private: 
    int r(vector<int>& nums) {
        int p1 = 0, p2 = 0, tmp = 0;
        for (int i = 0; i < nums.size(); ++i) {
            tmp = max(p2 + nums[i], p1);
            p2 = p1;
            p1 = tmp;
        }
        return tmp;
    }
public:
    int rob(vector<int>& nums) {
        if (nums.size() == 1) { return nums[0]; }
        vector<int> front(nums.begin(), nums.end() - 1);
        vector<int> back(nums.begin() + 1, nums.end());
        return max(r(front), r(back));
    }
};
